porch/controllers/packagevariants/pkg/controllers/packagevariant/packagevariant_controller.go

```diff
@@ -777,3 +782,158 @@ func getFileKubeObject(prr *porchapi.PackageRevisionResources, file, kind, name
 
 	return ko, nil
 }
+
+// TODO known issues:
+// - will move the newly generated name to the top of the mutator sequence
+// - does not preserve indent style. Instead will format if there are mutators/validators in pv, but will preserve if there are none.
+// - will not throw an error if the initial kptfile contains empty pipeline key
+func ensureKRMFunctions(pv *api.PackageVariant,
+	prr *porchapi.PackageRevisionResources) error {
+
+	if pv.Spec.Pipeline.IsEmpty() {
+		return nil
+	}
+
+	// parse kptfile
+	if _, ok := prr.Spec.Resources[kptfilev1.KptFileName]; !ok {
+		return fmt.Errorf("%s not found in PackageRevisionResources '%s/%s'", kptfilev1.KptFileName, prr.Namespace, prr.Name)
+	}
+	kptfile, err := fn.ParseKubeObject([]byte(prr.Spec.Resources[kptfilev1.KptFileName]))
+	if err != nil {
+		return fmt.Errorf("failed to parse %s of PackageRevisionResources '%s/%s'", kptfilev1.KptFileName, prr.Namespace, prr.Name)
+	}
+
+	pipeline := kptfile.UpsertMap("pipeline")
+
+	// generate new mutators
+	var newMutators = fn.SliceSubObjects{}
+
+	existingmutators, ok, err := pipeline.NestedSlice("mutators")
+	if err != nil {
+		return err
+	}
+	if !ok || existingmutators == nil {
+		existingmutators = fn.SliceSubObjects{}
+	}
+
+	for _, mutator := range existingmutators {
+		ok, err := isPackageVariantFunc(mutator, pv.ObjectMeta.Name)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			newMutators = append(newMutators, mutator)
+		}
+	}
+
+	var newPVMutators = fn.SliceSubObjects{}
+	for i, mutator := range pv.Spec.Pipeline.Mutators {
+		newMutator := mutator.DeepCopy()
+		newMutator.Name = generatePVFuncName(mutator.Name, pv.ObjectMeta.Name, i)
+		mut, err := fn.NewFromTypedObject(newMutator)
+		if err != nil {
+			return err
+		}
+		newPVMutators = append(newPVMutators, &mut.SubObject)
+	}
+
+	newMutators = append(newPVMutators, newMutators...)
+
+	// generate new validators
```

Comment:
I think you may be able to do mutators and validators in a  loop:
```
for _, field in range []string{"mutators", "validators"} {
...
}
```

That has its pros and cons, but seems do-able if you want to.

Reply:
I really like the idea, but I am unsure how to make it viable in Go. The line I am struggling to make work inside the loop is:

```Go
for i, mutator := range pv.Spec.Pipeline.Mutators {
```

Where a different field of pv.Spec.Pipeline should be taken depending on the value of the loop variable. The only solution I could really come up with would include a switch-case statement or some reflection, which I find rather ugly and not nice to maintain.

If you know of any elegant solution for this, I am very happy to change this part :)

Reply:
Well...not sure if it's *elegant*, but here's a solution:

```go
 funclists := map[string][]kptfilev1.Function{
        {"validators": pv.Spec.Pipeline.Validators},
        {"mutators": pv.Spec.Pipeline.Mutators},
}

for field, funclist := range funclists {
...do your thing...
}

```